File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxi = INT_MIN;
    int recur(TreeNode *root){
        if(root == NULL) return 0;
        int left = 0, right = 0;
        if(root->left) left = 1 + recur(root->left);
        if(root->right) right = 1 + recur(root->right);
        maxi = max(maxi, left+right);
        return max(left,right);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        recur(root);
        return maxi;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        # For each node, the path will be left + root + right.
        # return 1 + max(left, right)

        maxi = 0

        def recur(node):
            nonlocal maxi
            if node is None:
                return 0
            
            left = recur(node.left)
            right = recur(node.right)
            
            maxi = max(maxi, 1 + left + right)

            return 1 + max(left, right)
        
        recur(root)

        return maxi-1
        